base_layer/wallet/src/transaction_service/storage/models.rs

```diff
@@ -237,6 +244,117 @@ impl CompletedTransaction {
+    /// - `num_outputs`: Number of transaction outputs for the new transaction
+    ///
+    /// # Returns
+    /// A tuple of (weight_in_grams, fee_per_gram) where:
+    /// - `weight_in_grams`: The calculated transaction weight in grams
+    /// - `fee_per_gram`: The fee per gram calculated from total_fee / weight_in_grams
+    ///
+    /// # Notes
+    /// Calculates actual features and scripts sizes from the original transaction outputs,
+    /// then converts total fee to fee_per_gram using floating point division to avoid truncation
+    pub fn calculate_fee_per_gram_from_total_fee(
+        &self,
+        total_fee: MicroMinotari,
+        consensus_manager: &ConsensusManager,
+        num_inputs: usize,
+        num_outputs: usize,
+    ) -> Result<(u64, MicroMinotari), TransactionStorageError> {
+        let consensus_constants = consensus_manager.consensus_constants(0);
+        let fee_calculator = Fee::new(*consensus_constants.transaction_weight_params());
+
+        // Calculate average features and scripts size from actual transaction outputs
+        let original_outputs = self.transaction.body.outputs();
+        let total_features_and_scripts_size: Result<usize, std::io::Error> = original_outputs
+            .iter()
+            .map(|output| output.get_features_and_scripts_size())
+            .sum();
+
+        let total_features_and_scripts_size = total_features_and_scripts_size.map_err(|e| {
+            TransactionStorageError::FailedToCalculateTransactionFee(format!(
+                "Failed to calculate features and scripts size from original transaction: {}",
+                e
+            ))
+        })?;
+
+        // Calculate average size per output from original transaction
+        let avg_features_and_scripts_size_per_output = if original_outputs.is_empty() {
+            // Fallback to default if no outputs (shouldn't happen, but just in case)
+            let default_output_features_size = OutputFeatures::default().get_serialized_size().map_err(|e| {
+                TransactionStorageError::FailedToCalculateTransactionFee(format!(
+                    "Failed to serialize OutputFeatures: {}",
+                    e
+                ))
+            })?;
+            let default_script_size = TariScript::default().get_serialized_size().map_err(|e| {
+                TransactionStorageError::FailedToCalculateTransactionFee(format!(
+                    "Failed to serialize TariScript: {}",
+                    e
+                ))
+            })?;
+            let default_covenant_size = Covenant::new().get_serialized_size().map_err(|e| {
+                TransactionStorageError::FailedToCalculateTransactionFee(format!("Failed to serialize Covenant: {}", e))
+            })?;
+            default_output_features_size + default_script_size + default_covenant_size
+        } else {
+            total_features_and_scripts_size / original_outputs.len()
+        };
```

Comment:
The original transaction cannot have zero outputs, and zero length can be checked above as suggested, then we can simplify this.
```suggestion
        // Calculate average size per output from original transaction
        let avg_features_and_scripts_size_per_output = total_features_and_scripts_size / original_outputs.len();
```

Reply:
Changed it as suggested.